server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1201,34 +1204,158 @@ void removeEntitlement(Entitlement entitlement,
+            ent.getPool().getProvidedProducts().size();
+            Pool pool = ent.getPool();
+            pool.setConsumed(pool.getConsumed() - ent.getQuantity());
+
+            Consumer consumer = ent.getConsumer();
+            if (consumer.getType().isManifest()) {
+                pool.setExported(pool.getExported() - ent.getQuantity());
+            }
+        }
+
+        log.info("Starting batch delete of pools and entitlements");
+        poolCurator.batchDelete(poolsToDelete);
+        entitlementCurator.batchDelete(entsToRevoke);
+        entitlementCurator.flush();
+        log.info("All deletes flushed successfully");
+
+        List<Entitlement> stackingEntitlements = filterStackingEntitlements(entsToRevoke);
+
+        for (Entitlement ent : stackingEntitlements) {
+            Pool pool = ent.getPool();
+            Consumer consumer = ent.getConsumer();
+            String stackId = pool.getProductAttributeValue("stacking_id");
+            Pool stackedSubPool = poolCurator.getSubPoolForStackId(consumer, stackId);
+            if (stackedSubPool != null) {
+                List<Entitlement> stackedEnts =
+                    this.entitlementCurator.findByStackId(consumer, stackId);
+
+                // If there are no stacked entitlements, we need to delete the
+                // stacked sub pool, else we update it based on the entitlements
+                // currently in the stack.
+                if (stackedEnts.isEmpty()) {
+                    log.info("Revoke Entitlements: deleting single stacked sub-pool");
+                    deletePool(stackedSubPool);
+                }
+                else {
+                    updatePoolFromStackedEntitlements(stackedSubPool, stackedEnts);
+                    poolCurator.merge(stackedSubPool);
+                }
+            }
+        }
+
+        // post unbind actions
+        for (Entitlement ent : entsToRevoke) {
+            PoolHelper poolHelper = new PoolHelper(this, productCache, ent);
+            enforcer.postUnbind(ent.getConsumer(), poolHelper, ent);
+        }
+
+        List<Entitlement> batch = new ArrayList<Entitlement>();
+        for (int i = 0; i < entsToRevoke.size(); i++) {
+            Entitlement entitlement = entsToRevoke.get(i);
+            batch.add(entitlement);
+
+            // We work in batches of maximum size 1000.
+            if (i % 1000 == 0) {
+                Set<Entitlement> modifiedEnts = entitlementCurator.batchListModifying(batch);
+                this.regenerateCertificatesOf(modifiedEnts, true);
+                batch.clear();
+            }
+        }
+        log.info("Modifier entitlements done.");
+
+        if (!batch.isEmpty()) {
+            this.regenerateCertificatesOf(entitlementCurator.batchListModifying(batch), true);
+        }
+
+        Set<Consumer> distinctConsumers = entitlementCurator.getDistinctConsumers(entsToRevoke);
+
+        log.info("Recomputing status for " + distinctConsumers.size() + " consumers.");
```

Comment:
> You are suggesting to use log.isInfoEnabled()?

No, log.info (log.*level* generally) with SLF4J can take `{}` as place holders and will then populate them with subsequent arguments.  The place holder syntax is more concise and and log.info has a guard within it to immediately return if the log.level is higher than info.  Then end result is that toString() doesn't get called on the subsequent arguments.

So 

```
log.info("Value is " + someObject);
```

will result in a call to `someObject.toString()` before log.info is even called, since the arguments need to be resolved before they can be sent to the method.

```
log.info("Value is {}", someObject);
```

defers the cost of `toString()` until after the logger checks the level.  It's the equivalent of

```
if (log.isInfoEnabled()) {
  log.info("Value is " + someObject);
}
```